xds/csds/csds_test.go

```diff
@@ -0,0 +1,672 @@
+			return false
+		}
+		if b.EndpointConfig == nil {
+			return true
+		}
+		var at, bt v3endpointpb.ClusterLoadAssignment
+		if err := ptypes.UnmarshalAny(a.EndpointConfig, &at); err != nil {
+			panic("failed to unmarshal Endpoints" + err.Error())
+		}
+		if err := ptypes.UnmarshalAny(b.EndpointConfig, &bt); err != nil {
+			panic("failed to unmarshal Endpoints" + err.Error())
+		}
+		return strings.Compare(at.ClusterName, bt.ClusterName) < 0
+	}),
+	protocmp.IgnoreFields(&v3adminpb.ListenersConfigDump_DynamicListenerState{}, "last_updated"),
+	protocmp.IgnoreFields(&v3adminpb.RoutesConfigDump_DynamicRouteConfig{}, "last_updated"),
+	protocmp.IgnoreFields(&v3adminpb.ClustersConfigDump_DynamicCluster{}, "last_updated"),
+	protocmp.IgnoreFields(&v3adminpb.EndpointsConfigDump_DynamicEndpointConfig{}, "last_updated"),
+	protocmp.Transform(),
+}
+
+var (
+	ldsTargets   = []string{"lds.target.good:0000", "lds.target.good:1111"}
+	listeners    = make([]*v3listenerpb.Listener, len(ldsTargets))
+	listenerAnys = make([]*anypb.Any, len(ldsTargets))
+
+	rdsTargets = []string{"route-config-0", "route-config-1"}
+	routes     = make([]*v3routepb.RouteConfiguration, len(rdsTargets))
+	routeAnys  = make([]*anypb.Any, len(rdsTargets))
+
+	cdsTargets  = []string{"cluster-0", "cluster-1"}
+	clusters    = make([]*v3clusterpb.Cluster, len(cdsTargets))
+	clusterAnys = make([]*anypb.Any, len(cdsTargets))
+
+	edsTargets   = []string{"endpoints-0", "endpoints-1"}
+	endpoints    = make([]*v3endpointpb.ClusterLoadAssignment, len(edsTargets))
+	endpointAnys = make([]*anypb.Any, len(edsTargets))
+	ips          = []string{"0.0.0.0", "1.1.1.1"}
+	ports        = []uint32{123, 456}
+)
+
+func init() {
+	for i := range ldsTargets {
+		listeners[i] = e2e.DefaultListener(ldsTargets[i], rdsTargets[i])
+		listenerAnys[i], _ = ptypes.MarshalAny(listeners[i])
+	}
+	for i := range rdsTargets {
+		routes[i] = e2e.DefaultRouteConfig(rdsTargets[i], ldsTargets[i], cdsTargets[i])
+		routeAnys[i], _ = ptypes.MarshalAny(routes[i])
+	}
+	for i := range cdsTargets {
+		clusters[i] = e2e.DefaultCluster(cdsTargets[i], edsTargets[i])
+		clusterAnys[i], _ = ptypes.MarshalAny(clusters[i])
+	}
+	for i := range edsTargets {
+		endpoints[i] = e2e.DefaultEndpoint(edsTargets[i], ips[i], ports[i])
+		endpointAnys[i], _ = ptypes.MarshalAny(endpoints[i])
+	}
+}
+
+func TestCSDS(t *testing.T) {
+	const retryCount = 10
+
+	xdsC, mgmServer, nodeID, stream, cleanup := commonSetup(t)
+	defer cleanup()
+
+	for _, target := range ldsTargets {
+		xdsC.WatchListener(target, func(client.ListenerUpdate, error) {})
+	}
+	for _, target := range rdsTargets {
+		xdsC.WatchRouteConfig(target, func(client.RouteConfigUpdate, error) {})
+	}
+	for _, target := range cdsTargets {
+		xdsC.WatchCluster(target, func(client.ClusterUpdate, error) {})
+	}
+	for _, target := range edsTargets {
+		xdsC.WatchEndpoints(target, func(client.EndpointsUpdate, error) {})
+	}
+
+	for i := 0; i < retryCount; i++ {
```

Comment:
Is there something better than retrying in a loop and sleeping in between that we can do here?